### PR TITLE
🌱 Implement Machine Deleting condition

### PR DESCRIPTION
**Pull request by sbueringer**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #11105

### CODE REVIEW

File: internal/controllers/machine/machine_controller.go [hunk 1]
```diff
@@ -460,6 +503,8 @@ func (r *Reconciler) reconcileDelete(ctx context.Context, s *scope) (ctrl.Result
 			if conditions.Get(m, clusterv1.VolumeDetachSucceededCondition) == nil {
 				conditions.MarkFalse(m, clusterv1.VolumeDetachSucceededCondition, clusterv1.WaitingForVolumeDetachReason, clusterv1.ConditionSeverityInfo, "Waiting for node volumes to be detached")
 			}
+			s.deletingReason = clusterv1.MachineDeletingWaitingForVolumeDetachV1Beta2Reason
+			s.deletingMessage = ""
```
Comment (fabriziopandini) [inline internal/controllers/machine/machine_controller.go:507]: Am I wrong in assuming we will get more info in this message when @chrischdi PR merges?

File: internal/controllers/machine/machine_controller.go [hunk 2]
```diff
@@ -383,6 +401,25 @@ func (r *Reconciler) reconcileDelete(ctx context.Context, s *scope) (ctrl.Result
...
+			return ctrl.Result{RequeueAfter: requeueAfter}, nil
+		}
+	}
+
+	s.reconcileDeleteExecuted = true
+
+	// Add entry to the reconcileDeleteCache so we won't run reconcileDelete more than once per second.
+	// Under certain circumstances the ReconcileAfter time will be set to a later time, e.g. when we're waiting
+	// for Pods to terminate or volumes to detach.
+	// This is done to ensure we're not spamming the workload cluster API server.
```
Comment (fabriziopandini) [inline internal/controllers/machine/machine_controller.go:416]: If the goal is to protect the workload cluster, I will raise the interval to 3 or 5 seconds.

File: internal/controllers/machine/machine_controller.go [hunk 3]
```diff
@@ -383,6 +401,25 @@ func (r *Reconciler) reconcileDelete(ctx context.Context, s *scope) (ctrl.Result
...
+	r.reconcileDeleteCache.Add(cache.NewReconcileEntry(s.machine, time.Now().Add(1*time.Second)))
+
+	// Set "fallback" reason and message. This is used if we don't set a more specific reason and message below.
+	s.deletingReason = clusterv1.MachineDeletingDeletionTimestampSetV1Beta2Reason
```
Comment (fabriziopandini) [inline internal/controllers/machine/machine_controller.go:420]: q: what are the cases where we don't set a more specific reason?
I just want to double check that there are no conditions were deletions can be stuck without users getting additional info

File: internal/controllers/machine/machine_controller.go [hunk 4]
```diff
@@ -425,6 +464,8 @@ func (r *Reconciler) reconcileDelete(ctx context.Context, s *scope) (ctrl.Result
 			if conditions.Get(m, clusterv1.DrainingSucceededCondition) == nil {
 				conditions.MarkFalse(m, clusterv1.DrainingSucceededCondition, clusterv1.DrainingReason, clusterv1.ConditionSeverityInfo, "Draining the node before deletion")
 			}
+			s.deletingReason = clusterv1.MachineDeletingDrainingNodeV1Beta2Reason
+			s.deletingMessage = ""
```
Comment (fabriziopandini) [inline internal/controllers/machine/machine_controller.go:468]: What about 
```suggestion
			s.deletingMessage = fmt.Sprintf("Drain not completed yet (started at %s):", nodeDrainStartTime.Format(time.RFC3339))
```
(if there are chances that this will not be overridden down below)

File: internal/controllers/machine/machine_controller.go (hunk 1 above)
Comment (fabriziopandini) [reply internal/controllers/machine/machine_controller.go:507]: Also, let's add the start time here, in case there are code branches not setting the condition below

File: internal/controllers/machine/machine_controller_status.go [hunk 5]
```diff
@@ -461,15 +507,38 @@ func (c machineConditionCustomMergeStrategy) Merge(conditions []v1beta2condition
...
 	}).Merge(conditions, conditionTypes)
 }
 
+func setDeletingCondition(_ context.Context, machine *clusterv1.Machine, reconcileDeleteExecuted bool, deletingReason, deletingMessage string) {
+	if machine.DeletionTimestamp.IsZero() {
+		v1beta2conditions.Set(machine, metav1.Condition{
+			Type:   clusterv1.MachineDeletingV1Beta2Condition,
+			Status: metav1.ConditionFalse,
+			Reason: clusterv1.MachineDeletingDeletionTimestampNotSetV1Beta2Reason,
+		})
```
Comment (fabriziopandini) [inline internal/controllers/machine/machine_controller_status.go:520]: Missing return?

File: internal/controllers/machine/machine_controller_status.go [hunk 6]
```diff
@@ -499,6 +585,39 @@ func setReadyCondition(ctx context.Context, machine *clusterv1.Machine) {
 	v1beta2conditions.Set(machine, *readyCondition)
 }
 
+func calculateDeletingConditionForSummary(machine *clusterv1.Machine) v1beta2conditions.ConditionWithOwnerInfo {
```
Comment (fabriziopandini) [inline internal/controllers/machine/machine_controller_status.go:588]: Let's add a comment explaining why we compute a deleting condition for summary
(e.g. to drop granular details from the message

File: internal/controllers/machine/machine_controller_status.go [hunk 7]
```diff
@@ -499,6 +585,39 @@ func setReadyCondition(ctx context.Context, machine *clusterv1.Machine) {
...
+	deletingCondition := v1beta2conditions.Get(machine, clusterv1.MachineDeletingV1Beta2Condition)
+
+	var msg string
+	switch {
+	case deletingCondition == nil:
+		msg = "Machine deletion in progress"
```
Comment (fabriziopandini) [inline internal/controllers/machine/machine_controller_status.go:594]: Is this a case that should never happen? if yes, let's add a note

File: util/conditions/v1beta2/options.go [hunk 8]
```diff
@@ -79,6 +79,14 @@ func (t IgnoreTypesIfMissing) ApplyToSummary(opts *SummaryOptions) {
 	opts.ignoreTypesIfMissing = t
 }
 
+// AdditionalConditions allows to add additional conditions that should be used in the summary.
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply util/conditions/v1beta2/options.go:82]: Might be my previous comment is not correct and the intent here is to override conditions?

File: util/conditions/v1beta2/summary.go [hunk 9]
```diff
@@ -85,6 +87,12 @@ func NewSummaryCondition(sourceObj Getter, targetConditionType string, opts ...S
 		existingConditionTypes.Insert(condition.Type)
 	}
 
+	for _, condition := range summarizeOpt.additionalConditions {
+		conditionsInScope = append(conditionsInScope, condition)
+		existingConditionTypes.Insert(condition.Type)
+		allConditionTypes = append(allConditionTypes, condition.Type)
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply util/conditions/v1beta2/summary.go:93]: Might be my previous comment is not correct and the intent here is to override conditions?

File: api/v1beta1/machine_types.go [hunk 10]
```diff
@@ -271,6 +279,54 @@ const (
 const (
 	// MachineDeletingV1Beta2Condition surfaces details about progress in the machine deletion workflow.
 	MachineDeletingV1Beta2Condition = DeletingV1Beta2Condition
+
+	// MachineDeletingDeletingV1Beta2Reason surfaces when the Machine is deleting.
+	// This reason is only used for the MachineDeletingV1Beta2Condition when calculating the
+	// Ready condition when the deletionTimestamp on a Machine is set.
+	MachineDeletingDeletingV1Beta2Reason = DeletingV1Beta2Condition
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply api/v1beta1/machine_types.go:286]: > I know we are using mechanically condition + reason, but all those repetition are not easy to read. At least let's drop repeated words

Fine for me

> Also IMO we should not use a condition const to create a reason

That was not intentional, fixed

File: internal/controllers/machine/drain/drain.go [hunk 11]
```diff
@@ -301,12 +301,12 @@ func (r EvictionResult) DrainCompleted() bool {
...
 
 // ConditionMessage returns a condition message for the case where a drain is not completed.
-func (r EvictionResult) ConditionMessage() string {
+func (r EvictionResult) ConditionMessage(nodeDrainStartTime *metav1.Time) string {
 	if r.DrainCompleted() {
 		return ""
 	}
 
-	conditionMessage := "Drain not completed yet:"
+	conditionMessage := fmt.Sprintf("Drain not completed yet (started at %s):", nodeDrainStartTime.Format(time.RFC3339))
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/controllers/machine/drain/drain.go:309]: I think I am doing the same for volume detach

```go
s.deletingMessage = fmt.Sprintf("Waiting for Node volumes to be detached (started at %s)",
```

File: internal/util/cache/cache.go [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+	// Note: entries expire after the ttl.
+	Add(entry E)
+
+	// Has checks if the given key (still) exists in the Cache.
+	// Note: entries expire after the ttl.
+	Has(key string) (E, bool)
+}
+
+// New creates a new cache.
+func New[E Entry]() Cache[E] {
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/util/cache/cache.go:55]: I would prefer the current implementation that makes it explicit that the type used has to be able to calculate a unique key.

The Stringer interface does not give that guarantee

File: internal/util/cache/cache.go [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+	item, exists, _ := r.Store.GetByKey(key)
+	if exists {
+		return item.(E), true
+	}
+	return *new(E), false
+}
+
+// ReconcileEntry is an Entry for the Cache that stores the
+// earliest time after which the next Reconcile should be executed.
+type ReconcileEntry struct {
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/util/cache/cache.go:100]: See my comment above. Maybe let's go with the current implementation and evolve it later if it ever becomes necessary?

File: main.go [hunk 14]
```diff
@@ -178,6 +179,10 @@ func InitFlags(fs *pflag.FlagSet) {
 		"Grace period after which the RemoteConnectionProbe condition on a Cluster goes to false, "+
 			"if connection failures to a workload cluster occur")
 
+	fs.DurationVar(&remoteConditionsGracePeriod, "remote-conditions-grace-period", 5*time.Minute,
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply main.go:182]: Good point

File: main.go [hunk 15]
```diff
@@ -178,6 +179,10 @@ func InitFlags(fs *pflag.FlagSet) {
...
+		"Grace period after which remote conditions (e.g. `NodeHealthy`) are set to `Unknown`, "+
+			"if connection failures to a workload cluster occur")
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply main.go:184]: Done

File: util/conditions/v1beta2/options.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply util/conditions/v1beta2/options.go:82]: I think we usually would use it to overwrite an existing condition and not to make up an entirely new one.

I'm absolutely flexible regarding naming.

File: util/conditions/v1beta2/summary.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply util/conditions/v1beta2/summary.go:93]: > Let's add external conditions only if they do not already exists, so we prevent misconfigurations (we can consider also returning an error in this case, but in this case we should also change the for loop above).

I was playing a bit around with this. The problem is that if I would enforce that the same condition doesn't exist on the object I have to either:
* copy the object and remove the deleting condition
* use another condition type

File: internal/controllers/machine/machine_controller.go (hunk 1 above)
Comment (sbueringer) [reply internal/controllers/machine/machine_controller.go:507]: > Am I wrong in assuming we will get more info in this message when @chrischdi PR merges?

Not this one here, but I think the one in shouldWaitForNodeVolumes

> Also, let's add the start time here, in case there are code branches not setting the condition below

done

File: internal/controllers/machine/machine_controller.go (hunk 2 above)
Comment (sbueringer) [reply internal/controllers/machine/machine_controller.go:416]: The problem is that this increases the time for a happy path delete by probably 4-5 x (3s or 5s)

We also have client-side rate-limiting

File: internal/controllers/machine/machine_controller.go (hunk 3 above)
Comment (sbueringer) [reply internal/controllers/machine/machine_controller.go:420]: I'm mostly doing this because it's very hard to reason through all the branches below for all edge cases

File: internal/controllers/machine/machine_controller_status.go (hunk 5 above)
Comment (sbueringer) [reply internal/controllers/machine/machine_controller_status.go:520]: Thx!! great catch

Comment (sbueringer) [reply internal/controllers/machine/machine_controller_status.go:520]: Btw, it would still have worked because in that case reconcileDeleteExecuted is false (also the reason why the unit test didn't catch it)

File: internal/controllers/machine/machine_controller.go (hunk 3 above)
Comment (sbueringer) [reply internal/controllers/machine/machine_controller.go:420]: Audited the cases:
```go
return ctrl.Result{}, errors.Wrapf(err, "failed to check if Kubernetes Node deletion is allowed")
return ctrl.Result{}, err (l.460)
return ctrl.Result{}, errors.Wrap(err, "failed to patch Machine") (l.478)
return result, nil (l.491) because of return ctrl.Result{RequeueAfter: time.Minute}, nil (l.765)
return ctrl.Result{}, err (l.549, reconcileDeleteInfrastructure)
return ctrl.Result{}, err (l.561, reconcileDeleteBootstrap)
```

So I'll add deletingReason & deletingMessage for all of these as well.

Please double check, I'm absolutely not sure if I missed some cases

File: internal/controllers/machine/drain/drain.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (chrischdi) [reply internal/controllers/machine/drain/drain.go:309]: Doesn't this break grouping during aggregation? different machines have a different timestamp?

But we may not want to aggregate this message anyway because it usually is too verbose, right?

File: internal/controllers/machine/machine_controller_status.go [hunk 16]
```diff
@@ -499,6 +585,39 @@ func setReadyCondition(ctx context.Context, machine *clusterv1.Machine) {
...
+	case deletingCondition.Reason == clusterv1.MachineDeletingDrainingNodeV1Beta2Reason &&
+		machine.Status.Deletion != nil && machine.Status.Deletion.NodeDrainStartTime != nil &&
+		time.Since(machine.Status.Deletion.NodeDrainStartTime.Time) > 30*time.Minute:
+		msg = fmt.Sprintf("Machine deletion in progress, stage: %s (since more than 30m)", deletingCondition.Reason)
```
Comment (enxebre) [inline internal/controllers/machine/machine_controller_status.go:604]: why is 30m a threshold? why don't we ever print duration?

File: internal/controllers/clusterclass/clusterclass_controller.go [hunk 17]
```diff
@@ -73,7 +73,6 @@ func (r *Reconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, opt
 	predicateLog := ctrl.LoggerFrom(ctx).WithValues("controller", "clusterclass")
 	err := ctrl.NewControllerManagedBy(mgr).
 		For(&clusterv1.ClusterClass{}).
-		Named("clusterclass").
```
Comment (enxebre) [inline internal/controllers/clusterclass/clusterclass_controller.go:76]: this is unrelated I assume?

File: internal/controllers/machine/drain/drain.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/controllers/machine/drain/drain.go:309]: @chrischdi Are we aggregating Deleting conditions by message?

File: internal/controllers/machine/machine_controller_status.go [hunk 18]
```diff
@@ -66,12 +70,12 @@ func (r *Reconciler) reconcileStatus(ctx context.Context, s *scope) {
 
 	// TODO: Set the uptodate condition for standalone pods
 
+	setDeletingCondition(ctx, s.machine, s.reconcileDeleteExecuted, s.deletingReason, s.deletingMessage)
```
Comment (enxebre) [inline internal/controllers/machine/machine_controller_status.go:73]: Stefan Instead of having scope variables that mutate through reconciliation have you considered putting the checkConditionStatus logic in a centralized single place? That might result in some code/execution duplication but it would make easy to grasp and maintain status. On the other side it might not always be practical so just curious if you thought about it

File: internal/controllers/clusterclass/clusterclass_controller.go (hunk 17 above)
Comment (sbueringer) [reply internal/controllers/clusterclass/clusterclass_controller.go:76]: Yeah absolutely unrelated. Just saw it when doing some research around controller names while I had this branch open. I can open a separate PR if necessary

File: internal/controllers/machine/machine_controller_status.go (hunk 18 above)
Comment (sbueringer) [reply internal/controllers/machine/machine_controller_status.go:73]: Do you mean putting all the code needed to calculate the condition into a single place? The problem is that we need a huge part of our controller code to caculate the conditions. So we either would have to duplicate these parts or move them all into this single place

(I might have misunderstood what you're suggesting though)

File: util/conditions/v1beta2/options.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply util/conditions/v1beta2/options.go:82]: Let's use override then, so we are explicit about the use case we support

File: util/conditions/v1beta2/summary.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply util/conditions/v1beta2/summary.go:93]: Since we are shifting to override, let's do the other way around
OverrideConditions must exists if not in ignoreTypesIfMissing, otherwise error

File: internal/controllers/machine/machine_controller.go (hunk 3 above)
Comment (fabriziopandini) [reply internal/controllers/machine/machine_controller.go:420]: I'll defer this to you
I'm ok if we skip the less probable cases (where e.g. we read only stuff from the cache, or we perform simple operations on the management cluster)

File: internal/util/cache/cache.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply internal/util/cache/cache.go:55]: ack

File: internal/controllers/machine/drain/drain.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply internal/controllers/machine/drain/drain.go:309]: Aren't we getting around this by bucketing stuff in less than 30m / more than 30m?

File: util/conditions/v1beta2/options.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply util/conditions/v1beta2/options.go:82]: Should I only rename the field or should I also fail if the sourceObject doesn't have the condition?

File: internal/controllers/machine/drain/drain.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/controllers/machine/drain/drain.go:309]: This is the Deleting condition. Not the message we surface in the Ready condition

File: util/conditions/v1beta2/options.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply util/conditions/v1beta2/options.go:82]: Just read your comment below

File: internal/controllers/machine/machine_controller.go (hunk 3 above)
Comment (sbueringer) [reply internal/controllers/machine/machine_controller.go:420]: I added reasons & message for all of these cases. It would be just good if someone else also goes through all branches to ensure I didn't miss any. It's non-trivial :)

File: internal/controllers/machine/drain/drain.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply internal/controllers/machine/drain/drain.go:309]: Sorry, I wasn't clear.
IMO It is ok to have a detailed Deleting condition message, because we are never going to use it for aggreagation.
For aggregation we use a simplified message bucketing machines in two categories (less than 30m / more than 30m)

[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/controllers/machine/drain/drain.go:309]: To give an example.
```
Drain not completed yet (started at 2024-10-09T16:13:59Z):
* Pods with deletionTimestamp that still exist: pod-2-deletionTimestamp-set-1, pod-2-deletionTimestamp-set-2, pod-2-deletionTimestamp-set-3, pod-3-to-trigger-eviction-successfully-1, pod-3-to-trigger-eviction-successfully-2, ... (2 more)
* Pods with eviction failed:
  * Cannot evict pod as it would violate the pod's disruption budget. The disruption budget pod-5-pdb needs 20 healthy pods and has 20 currently: pod-5-to-trigger-eviction-pdb-violated-1, pod-5-to-trigger-eviction-pdb-violated-2, pod-5-to-trigger-eviction-pdb-violated-3, ... (3 more)
  * some other error 1: pod-6-to-trigger-eviction-some-other-error
  * some other error 2: pod-7-to-trigger-eviction-some-other-error
  * some other error 3: pod-8-to-trigger-eviction-some-other-error
  * some other error 4: pod-9-to-trigger-eviction-some-other-error
  * ... (1 more error applying to 1 Pod)`
```

It's very very unlikely that this will every look the same for different Machines, even if I get rid of the start time entirely)

File: internal/controllers/machine/machine_controller_status.go (hunk 18 above)
Comment (enxebre) [reply internal/controllers/machine/machine_controller_status.go:73]: Yes, that's what I meant. It'd be kind of inverting the loop logic, e.g. First calculate status by observing most recent state of the world (querying the world). Then reconcile intent and use the observed state to determine course of actions if needed. But anyways as mentioned it might be impractical and I don't want to hijack this PR on this discussion :) just something to think about.

File: util/conditions/v1beta2/summary.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply util/conditions/v1beta2/summary.go:93]: Refactored a bit. I think now it behaves as desired

File: util/conditions/v1beta2/options.go [hunk 19]
```diff
@@ -79,6 +79,14 @@ func (t IgnoreTypesIfMissing) ApplyToSummary(opts *SummaryOptions) {
 	opts.ignoreTypesIfMissing = t
 }
 
+// OverrideConditions allows to override conditions from the source object that should be used in the summary.
```
Comment (fabriziopandini) [inline util/conditions/v1beta2/options.go:82]:
```suggestion
// OverrideConditions allows to override conditions read from the source object only for the scope of a summary operation.
// The condition on the source object will preserve the original value.
```

File: util/conditions/v1beta2/summary.go [hunk 20]
```diff
@@ -74,14 +75,34 @@ func NewSummaryCondition(sourceObj Getter, targetConditionType string, opts ...S
...
-	// Drops all the conditions not in scope for the merge operation
 	conditions := getConditionsWithOwnerInfo(sourceObj)
+
+	conditionsByType := map[string]ConditionWithOwnerInfo{}
+	for _, c := range conditions {
+		conditionsByType[c.Type] = c
+	}
+	overrideConditionsByType := map[string]ConditionWithOwnerInfo{}
+	for _, c := range summarizeOpt.overrideConditions {
+		overrideConditionsByType[c.Type] = c
```
Comment (fabriziopandini) [inline util/conditions/v1beta2/summary.go:86]: There is still the edge case when an override condition is passed more than one time, but I'm also ok with "the last value wins" as it is now.
Might be let's add a note on the option's go doc.

File: internal/controllers/machine/machine_controller_status.go [hunk 21]
```diff
@@ -499,6 +586,49 @@ func setReadyCondition(ctx context.Context, machine *clusterv1.Machine) {
...
 
+// calculateDeletingConditionForSummary calculates a Deleting condition for the calculation of the Ready condition
+// (which is done via a summary). This is necessary to avoid including the verbose details of the Deleting condition
+// message in the summary.
+// This is important to ensure we have a limited amount of unique messages across Machines. This allows us to deduplicate
+// messages when aggregating Ready conditions of many Machines into the MachinesReady condition of e.g. the MachineSet.
+// For the same reason we are only surfacing messages with "more than 30m" instead of using the exact durations.
+// We assume that 30m is a duration after which it makes sense to notify users that Node drains and waiting for volume
+// detach are still in progress.
+func calculateDeletingConditionForSummary(machine *clusterv1.Machine) v1beta2conditions.ConditionWithOwnerInfo {
```
Comment (fabriziopandini) [inline internal/controllers/machine/machine_controller_status.go:597]: nit
```suggestion
// calculateDeletingConditionForSummary calculates a Deleting condition for the calculation of the Ready condition
// (which is done via a summary). This is necessary to avoid including the verbose details of the Deleting condition
// message in the summary.
// This is also important to ensure we have a limited amount of unique messages across Machines thus allowing to
// nicely aggregate Ready conditions from many Machines into the MachinesReady condition of e.g. the MachineSet.
// For the same reason we are only surfacing messages with "more than 30m" instead of using the exact durations.
// "30m" is a an arbitrary duration after which we assume it makes sense to give more emphasis on 
the fact that Node drains and waiting for volume
// detach are still in progress. 
func computeDeletingOverrideConditionForSummary(machine *clusterv1.Machine) v1beta2conditions.ConditionWithOwnerInfo {
```

File: util/conditions/v1beta2/summary.go (hunk 20 above)
Comment (sbueringer) [reply util/conditions/v1beta2/summary.go:86]: Added a check and a unit test

File: internal/controllers/machine/machine_controller_status.go (hunk 21 above)
Comment (sbueringer) [reply internal/controllers/machine/machine_controller_status.go:597]: ~ changed to this. I didn't use arbitrary because it would mean we randomly picked this value, but we chose it intentionally.

**Comment by sbueringer** [comment]
Thx for the reviews!! Findings should be addressed or answered, PTAL :)

File: internal/controllers/machine/machine_controller_status.go [hunk 22]
```diff
@@ -213,8 +218,49 @@ func setInfrastructureReadyCondition(_ context.Context, machine *clusterv1.Machi
...
+			Type:    clusterv1.MachineNodeHealthyV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  clusterv1.MachineNodeRemoteConnectionFailedV1Beta2Reason,
+			Message: fmt.Sprintf("Remote connection down since %s", lastProbeSuccessTime.Format(time.RFC3339)),
+		})
+		return
+	}
+
+	if nodeGetErr != nil {
+		if errors.Is(nodeGetErr, clustercache.ErrClusterNotConnected) {
```
Comment (chrischdi) [inline internal/controllers/machine/machine_controller_status.go:240]: Should we set the condition to unknown if it does not exist yet? (to ensure the condition always exists)

**Comment by chrischdi** [comment]
Just one comment, otherwise lgtm :-)

File: internal/controllers/machine/machine_controller.go (hunk 3 above)
Comment (fabriziopandini) [reply internal/controllers/machine/machine_controller.go:419]: I double checked the code paths, it seems ok to me

File: internal/controllers/machine/machine_controller_status.go (hunk 22 above)
Comment (sbueringer) [reply internal/controllers/machine/machine_controller_status.go:240]: Done

**Comment by sbueringer** [comment]
@fabriziopandini @chrischdi @enxebre 
All findings should be addressed, PTAL :)
